File: go/verifier/attenuate.go
package verifier

import (
	"encoding/json"
	"reflect"
	"strings"
)

func globPrefix(resource string) (string, bool) {
	if strings.HasSuffix(resource, "*") {
		return resource[:len(resource)-1], true
	}
	return "", false
}

// ResourceContains reports whether the parent resource pattern covers the child. Only a trailing "*" is a glob.
func ResourceContains(parent, child string) bool {
	if parent == child {
		return true
	}
	pp, ok := globPrefix(parent)
	if !ok {
		return false
	}
	cp, cok := globPrefix(child)
	if !cok {
		cp = child
	}
	return strings.HasPrefix(cp, pp)
}

func asFloat(v any) (float64, bool) {
	switch n := v.(type) {
	case float64:
		return n, true
	case float32:
		return float64(n), true
	case int:
		return float64(n), true
	case int64:
		return float64(n), true
	case json.Number:
		f, err := n.Float64()
		return f, err == nil
	}
	return 0, false
}

func asStrings(v any) ([]string, bool) {
	switch s := v.(type) {
	case []string:
		return s, true
	case []any:
		out := make([]string, 0, len(s))
		for _, x := range s {
			str, ok := x.(string)
			if !ok {
				return nil, false
			}
			out = append(out, str)
		}
		return out, true
	}
	return nil, false
}

// ConstraintsTighter: are the child's constraints at least as tight as the parent's?
func ConstraintsTighter(parent, child Constraints) bool {
	if parent == nil {
		return true
	}
	if child == nil {
		child = Constraints{}
	}
	for key, p := range parent {
		if p == nil {
			continue
		}
		switch key {
		case "max_value", "max_total":
			pv, _ := asFloat(p)
			cv, ok := asFloat(child[key])
			if !ok || !(cv <= pv) {
				return false
			}
		case "region":
			pr, _ := asStrings(p)
			set := map[string]bool{}
			for _, r := range pr {
				set[r] = true
			}
			cr, ok := asStrings(child["region"])
			if !ok || len(cr) == 0 {
				return false
			}
			for _, r := range cr {
				if !set[r] {
					return false
				}
			}
		default:
			if !jsonEqual(child[key], p) {
				return false
			}
		}
	}
	return true
}

func jsonEqual(a, b any) bool {
	if reflect.DeepEqual(a, b) {
		return true
	}
	ja, ea := json.Marshal(a)
	jb, eb := json.Marshal(b)
	return ea == nil && eb == nil && string(ja) == string(jb)
}

// CapabilityCovered: is child covered by parent?
func CapabilityCovered(parent, child Capability) bool {
	return parent.Action == child.Action && ResourceContains(parent.Resource, child.Resource) && ConstraintsTighter(parent.Constraints, child.Constraints)
}

// IsSubset: every child capability is covered by some parent capability.
func IsSubset(child, parent CapSet) bool {
	for _, c := range child {
		covered := false
		for _, p := range parent {
			if CapabilityCovered(p, c) {
				covered = true
				break
			}
		}
		if !covered {
			return false
		}
	}
	return true
}

// Attenuate keeps only the requested capabilities the parent covers.
func Attenuate(parent, requested CapSet) CapSet {
	out := CapSet{}
	for _, c := range requested {
		for _, p := range parent {
			if CapabilityCovered(p, c) {
				out = append(out, c)
				break
			}
		}
	}
	return out
}

File: packages/verifier/src/delegation/attenuate.ts
import type { CapSet, Capability, Constraints } from "../types";

/** "supplier:*" → "supplier:", exact resources → null. Only a trailing "*" is a glob. */
function globPrefix(resource: string): string | null {
  return resource.endsWith("*") ? resource.slice(0, -1) : null;
}

/** Does the parent resource pattern cover the child pattern? */
export function resourceContains(parent: string, child: string): boolean {
  if (parent === child) return true;
  const pp = globPrefix(parent);
  if (pp === null) return false;                       // an exact parent covers only itself
  const cp = globPrefix(child);
  return (cp ?? child).startsWith(pp);                 // child (or its glob prefix) sits under the parent prefix
}

/**
 * Are the child's constraints at least as tight as the parent's?
 * - numeric limits (max_value): child must carry one, and it must not exceed the parent's
 * - currency: must match when the parent fixes it
 * - region: child set must be a subset when the parent restricts it
 * - any other parent key: child must carry an identical value (conservative)
 * The child may add constraint keys the parent lacks (that only tightens).
 */
export function constraintsTighter(parent: Constraints | undefined, child: Constraints | undefined): boolean {
  if (!parent) return true;
  const c = child ?? {};
  for (const key of Object.keys(parent)) {
    const p = parent[key];
    if (p === undefined) continue;
    if (key === "max_value" || key === "max_total") {
      const cv = c[key];
      if (typeof cv !== "number" || !(cv <= (p as number))) return false;
    } else if (key === "region") {
      const pr = new Set(p as string[]);
      const cr = c.region;
      if (!Array.isArray(cr) || cr.length === 0 || !cr.every((r) => pr.has(r))) return false;
    } else if (JSON.stringify(c[key]) !== JSON.stringify(p)) {
      return false;
    }
  }
  return true;
}

/** Is `child` capability covered by `parent` capability? */
export function capabilityCovered(parent: Capability, child: Capability): boolean {
  return parent.action === child.action
    && resourceContains(parent.resource, child.resource)
    && constraintsTighter(parent.constraints, child.constraints);
}

/** Every child capability is covered by some parent capability. */
export function isSubset(child: CapSet, parent: CapSet): boolean {
  return child.every((c) => parent.some((p) => capabilityCovered(p, c)));
}

/**
 * Attenuate: keep only the requested capabilities the parent actually covers.
 * By construction `isSubset(attenuate(parent, requested), parent)` always holds,
 * and the result never contains anything absent from `requested`.
 */
export function attenuate(parent: CapSet, requested: CapSet): CapSet {
  return requested.filter((c) => parent.some((p) => capabilityCovered(p, c)));
}
